Cryptic error when `package.json` is missing a name

**Is your feature request related to a problem? Please describe.**
I was trying out the experimental JavaScript backend and running some "unfriendly" behavior. I've included a minimal reproduction [repo](https://github.com/yjabri/pants-no-package-json-name) to try the following in.

Run `pants tailor ::`. It populates `foo/BUILD` with 

```

package_json()
```

Run `pants tailor ::` again. You'll get

```
$ pants tailor ::
14:56:17.68 [ERROR] 1 Exception encountered:

Engine traceback:
  in `tailor` goal

KeyError: 'name'
```

In a non minimal repo I had `package_json()` populate like `package_json(name="foo0")` since I had a python distribution with the same name. Given `KeyError: 'name'` it feels like `tailor` is writing an invalid target to the BUILD file. Also running `pants help targets package_json` doesn't enumerate a `name` field.

It took me some time to realize this had nothing to do with the BUILD file but instead that my `package.json` was missing a name.

**Describe the solution you'd like**
A more explicit message should be given to the caller that `package.json` needs to have a name the first time `pants tailor` (or any goal that consumes the file) is called.

**Describe alternatives you've considered**
-

**Additional context**
-

Thanks for the reproducer. I can reproduce. 

---

For someone debugging, here's the stack trace output `pants --print-stacktrace tailor ::`

```
09:44:13.39 [ERROR] 1 Exception encountered:

Engine traceback:
  in root
    ..
  in pants.core.goals.tailor.tailor
    `tailor` goal

Traceback (most recent call last):
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/core/goals/tailor.py", line 650, in tailor
    specs_paths = await Get(SpecsPaths, Specs, specs)
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 124, in __await__
    result = yield self
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/specs_rules.py", line 313, in resolve_specs_paths
    unfiltered_include_targets, ignore_targets, include_paths, ignore_paths = await MultiGet(
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 436, in MultiGet
    return await _MultiGet((__arg0, __arg1, __arg2, __arg3))
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 174, in __await__
    result = yield self.gets
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/specs_rules.py", line 258, in resolve_addresses_from_raw_specs
    without_file_owners, with_file_owners = await MultiGet(
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 406, in MultiGet
    return await _MultiGet((__arg0, __arg1))
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 174, in __await__
    result = yield self.gets
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/specs_rules.py", line 183, in addresses_from_raw_specs_without_file_owners
    target_parametrizations_list = await MultiGet(
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 376, in MultiGet
    return await _MultiGet(tuple(__arg0))
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 174, in __await__
    result = yield self.gets
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/graph.py", line 349, in resolve_target_parametrizations
    all_generated = await MultiGet(
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 376, in MultiGet
    return await _MultiGet(tuple(__arg0))
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 174, in __await__
    result = yield self.gets
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/backend/javascript/package_json.py", line 735, in all_package_json
    await Get(
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 124, in __await__
    result = yield self
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/backend/javascript/package_json.py", line 709, in read_package_jsons
    await MultiGet(
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 376, in MultiGet
    return await _MultiGet(tuple(__arg0))
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/engine/internals/selectors.py", line 174, in __await__
    result = yield self.gets
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/backend/javascript/package_json.py", line 688, in parse_package_json
    name=parsed_package_json["name"],
  File "/Users/huon/Library/Caches/nce/f46d2c12132bad9c27e0dd509186c190371abfe1add7e5ca42245466f35bea81/bindings/venvs/2.20.0/lib/python3.9/site-packages/pants/util/frozendict.py", line 77, in __getitem__
    return self._data[k]
KeyError: 'name'
```

--- a/docs/notes/2.23.x.md
+++ b/docs/notes/2.23.x.md
@@ -123,8 +123,8 @@ pants to invoke `corepack`s default versions of the package managers instead.
 (npm, pnpm, yarn) without providing a corresponding `[nodejs].package_managers` version setting. The version is then
 entirely handled by `corepack`. Previously this mode caused pants to fail.
 
-The internal installation mechanism for node_modules has changed. 
-Previously, Pants installed each package separately in sandboxes and merged the results, creating a node_modules for all dependent packages in the workspace. 
+The internal installation mechanism for node_modules has changed.
+Previously, Pants installed each package separately in sandboxes and merged the results, creating a node_modules for all dependent packages in the workspace.
 Now, this is delegated to the package managers, using each package manager's support for workspaces.
 
 This fixes an issue with integrity file collisions when newer versions of package managers (e.g. the [hidden lockfiles](https://docs.npmjs.com/cli/v9/configuring-npm/package-lock-json#hidden-lockfiles) introduced in npm v7).
@@ -132,6 +132,7 @@ This fixes an issue with integrity file collisions when newer versions of packag
 `pants export --resolve=<js-resolve>` now has basic support for exporting the package manager installation artifacts
 including `node_modules`. This can be used to inspect the installation, or to enable IDE:s to discover the packages.
 
+Pants will output a more helpful error message if there is no `name` field defined in the `package.json` file, or if the `name` field is empty.
 
 #### Shell
 

--- a/src/python/pants/backend/javascript/package_json.py
+++ b/src/python/pants/backend/javascript/package_json.py
@@ -683,9 +683,13 @@ async def find_owning_package(request: OwningNodePackageRequest) -> OwningNodePa
 @rule
 async def parse_package_json(content: FileContent) -> PackageJson:
     parsed_package_json = FrozenDict.deep_freeze(json.loads(content.content))
+    package_name = parsed_package_json.get("name")
+    if not package_name:
+        raise ValueError("No package name found in package.json")
+
     return PackageJson(
         content=parsed_package_json,
-        name=parsed_package_json["name"],
+        name=package_name,
         version=parsed_package_json.get("version"),
         snapshot=await Get(Snapshot, PathGlobs([content.path])),
         module=parsed_package_json.get("type"),

--- a/src/python/pants/backend/javascript/package_json_test.py
+++ b/src/python/pants/backend/javascript/package_json_test.py
@@ -27,7 +27,7 @@ from pants.engine.internals.native_engine import Snapshot
 from pants.engine.internals.scheduler import ExecutionError
 from pants.engine.rules import QueryRule
 from pants.engine.target import AllTargets
-from pants.testutil.rule_runner import RuleRunner
+from pants.testutil.rule_runner import RuleRunner, engine_error
 from pants.util.frozendict import FrozenDict
 
 
@@ -84,6 +84,18 @@ def test_parses_package_jsons(rule_runner: RuleRunner) -> None:
             snapshot=bar_package_snapshot,
         ),
     }
+
+
+def test_parse_package_json_without_name(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/js/BUILD": "package_json()",
+            # No name in package.json, should cause an error.
+            "src/js/package.json": json.dumps({"version": "0.0.1"}),
+        }
+    )
+    with engine_error(ValueError, contains="No package name found in package.json"):
+        rule_runner.request(AllPackageJson, [])
 
 
 def test_generates_third_party_node_package_targets(rule_runner: RuleRunner) -> None:

